### PR TITLE
Allow switch of importation reduction method

### PR DESCRIPTION
**Pull request by mattg3004**
In the simulation we want to reduce the importation rate to represent reality in some way, as a high importation rate becomes unrealistic as prevalence reduces, as this implies that in every place outside of the IU we are simulating the prevalence is unchanged. 

There are 2 ways in which we can do this:
1. Specify the reduction in the scenario file. If this is done then the way the reduction is calculated is via the data given in the maps for fitting the model.
2. When an MDA is done, check the prevalence 6 months later and reduce the importation rate by the ratio of the prevalence now divided by the prevalence before the MDA.

If we are doing case 1., then once we reach the present day (or the end of the map data), we no longer have any information to reduce the importation rate. This PR will allow us to switch to method 2. at a specified time within the simulation.

### CODE REVIEW

File: src/Model.cpp [hunk 1]
```diff
@@ -155,6 +155,18 @@ void Model::runScenarios(ScenariosList &scenarios, Population &popln,
   // finished
 }
 
+bool Model::reduceImportationViaPrevalenceCheck(
```
Comment (thk123) [inline src/Model.cpp:158]: ⛏️ For functions that return booleans, it is good if the function name reads like a question that the boolean is the answer to, so in this case `shouldReduceImportationViaPrevalance`

File: tests/CMakeLists.txt [hunk 2]
```diff
@@ -1,20 +1,21 @@
...
+          CMAKE_CURRENT_SOURCE_DIR
+        } "*.cpp") list(SORT ALL_FILES)
 
-# Check globbing doesn't find any other tests, whilst still ensuring that generated Make files rebuild appropriately. 
-if (NOT ("${ALL_FILES}" STREQUAL "${TESTS_TO_RUN}"))
-    message(INFO " Expected tests = ${TESTS_TO_RUN}")
-    message(INFO " Found tests = ${ALL_FILES}")
-    message(SEND_ERROR "Some tests are missing from TESTS_TO_RUN, please add in tests/CMakeLists.txt")
-endif()
+#Check globbing doesn't find any other tests, whilst still ensuring that generated Make files rebuild appropriately.
```
Comment (thk123) [inline tests/CMakeLists.txt]: ❓ Hmm what is going on with these formatting changes - are they intentional?

File: tests/test_model.cpp [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    t = 270;
+    REQUIRE(model.reduceImportationViaPrevalenceCheck(
+        reduceImpViaXml, t, switchImportationReducingMethodTime));
+
+    // the next two times the function is called should be false. This is
+    // because the reduceImpViaXml variable is set to 1 and the time is before
+    // switchImportationReducingMethodTime
+    reduceImpViaXml = 1;
+    t = 0;
+    REQUIRE(!model.reduceImportationViaPrevalenceCheck(
```
Comment (thk123) [inline tests/test_model.cpp:27]: 💡 You can use `REQUIRE_FALSE` to make it a little clearer (the `!` can be easy to miss!)

File: src/Model.cpp [hunk 4]
```diff
@@ -155,6 +155,18 @@ void Model::runScenarios(ScenariosList &scenarios, Population &popln,
...
+    int reduceImpViaXml, int t, int switchImportationReducingMethodTime) {
```
Comment (thk123) [inline src/Model.cpp:159]: ⛏️ While `reduceImpViaXml` was a `int` it should really be a `bool` (right?), so to avoid making the code worse, I'd have this function take a boolean, and when you're calling it do:

```C++
reduceImportationViaPrevalanceCheck(reduceImpViaXml == 1, ...
```

Then when someone tidies up the reduceImpViaXml elsewhere, they will have one less job to do!

File: tests/test_model.cpp [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+#include <catch2/catch_all.hpp>
+TEST_CASE("Model", "[classic]") {
+  SECTION("Model::reduceImportationViaPrevalenceCheck") {
+
+    Model model;
+
+    // test the model.reduceImportationViaPrevalenceCheck function
+    // this should return false if reduceImpViaXml = 1 and the time is below
+    // switchImportationReducingMethodTime. Otherwise this should be true
+    int reduceImpViaXml = 0;
```
Comment (thk123) [inline tests/test_model.cpp:11]: 💡 I would inline all of these values so you just have:
```
REQUIRE(model.reduceImportationViaPrevalenceCheck(0, 0, 270));
...
REQUIRE(model.reduceImportationViaPrevalenceCheck(0, 270, 270));
```

I would also do each test case as its own section, and in this section description you can put the expected behaviour (rather than in a comment) - this way if the test fails, the output will be a lot more useful:


```
Model model;
SECTION("Not reduced via xml, time before switch point, should do reduction") {
  REQUIRE(model.reduceImportationViaPrevalenceCheck(0, 0, 270));
}
SECTION("Not reduced via xml, time at the switch point, should do reduction") {
  REQUIRE(model.reduceImportationViaPrevalenceCheck(0, 270, 270));
}
```
Which to me doesn't seem right actually having written it out in plain English? ❓

**Comment by thk123** [comment]
Just some suggestions in my review - but the question in the test would be good to know the function is indeed doing the correct thing!

File: src/Model.cpp (hunk 1 above)
Comment (mattg3004) [reply src/Model.cpp:158]: changed function name to suggested

File: tests/test_model.cpp (hunk 3 above)
Comment (mattg3004) [reply tests/test_model.cpp:27]: edited this

**Review by thk123 — APPROVED** [review]
💚 The text in the text is most helpful and makes the function clearly correct!

**Comment by thk123** [comment]
Approved on the assumption you will fix whatever the reason for the build to fail is.
